### DISCUSSION TITLE
Tests and samples

### DISCUSSION THREAD

**Discussion by dmitry-kazakov** [alire-project/alire#1186] [Q&A]
What is the way of dealing with tests and samples to the crate? Are automated tests supported? I.e. tests built and even ran during installation.

**Comment by mosteo** [comment]
The current recommendation is to use a nested crate: https://alire.ada.dev/docs/#work-in-progress-dependency-overrides (look for the pins to local crates)

This isn't currently stated explicitly there (thought we should do it somewhere else I think) but the idea is that the nested crates just use the parent crate as a local dependency. Interested people can `alr get` the main crate and play with these tests/samples.

There are hooks for automatic actions (look for `[[actions]]` in https://github.com/alire-project/alire/blob/master/doc/catalog-format-spec.md#release-information) but we don't really want tests to be run during normal use of a lib as a dependency, so the current advice would be to have a nested testing crate that the user can `alr build` on-demand.

You'll see in that link that you can have test actions too. These are run during our internal testing and during release submission to the index. Conceivably, we can add an option for people interested in running tests during installation (which is not currently a feature). 

`alr test` is still a little green and doesn't allow to test your local crate yet, but if it eventually does, it will execute test actions just like now it does for remote crates. And these can dispatch to nested crates, of course, so no need to duplicate testing or bring testing dependencies into the main crate.
